src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Decoder.cs

```diff
@@ -1234,25 +1237,283 @@ internal static bool IsWhiteSpace(int value)
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            [CompExactlyDependsOn(typeof(AdvSimd.Arm64))]
+            [CompExactlyDependsOn(typeof(Ssse3))]
+            public static bool TryDecode128Core(
+                Vector128<byte> str,
+                Vector128<byte> hiNibbles,
+                Vector128<byte> maskSlashOrUnderscore,
+                Vector128<byte> mask8F,
+                Vector128<byte> lutLow,
+                Vector128<byte> lutHigh,
+                Vector128<sbyte> lutShift,
+                Vector128<byte> _,
+                out Vector128<byte> result)
+            {
+                Vector128<byte> loNibbles = str & maskSlashOrUnderscore;
+                Vector128<byte> hi = SimdShuffle(lutHigh, hiNibbles, mask8F);
+                Vector128<byte> lo = SimdShuffle(lutLow, loNibbles, mask8F);
+
+                // Check for invalid input: if any "and" values from lo and hi are not zero,
+                // fall back on bytewise code to do error checking and reporting:
+                if ((lo & hi) != Vector128<byte>.Zero)
+                {
+                    result = default;
+                    return false;
+                }
+
+                Vector128<byte> eq2F = Vector128.Equals(str, maskSlashOrUnderscore);
+                Vector128<byte> shift = SimdShuffle(lutShift.AsByte(), (eq2F + hiNibbles), mask8F);
+
+                // Now simply add the delta values to the input:
+                result = str + shift;
+
+                return true;
+            }
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            [CompExactlyDependsOn(typeof(Avx2))]
+            public static bool TryDecode256Core(
+                Vector256<sbyte> str,
+                Vector256<sbyte> hiNibbles,
+                Vector256<sbyte> maskSlashOrUnderscore,
+                Vector256<sbyte> lutLow,
+                Vector256<sbyte> lutHigh,
+                Vector256<sbyte> lutShift,
+                Vector256<sbyte> _,
+                out Vector256<sbyte> result)
+            {
+                Vector256<sbyte> loNibbles = Avx2.And(str, maskSlashOrUnderscore);
+                Vector256<sbyte> hi = Avx2.Shuffle(lutHigh, hiNibbles);
+                Vector256<sbyte> lo = Avx2.Shuffle(lutLow, loNibbles);
+
+                if (!Avx.TestZ(lo, hi))
+                {
+                    result = default;
+                    return false;
+                }
+
+                Vector256<sbyte> eq2F = Avx2.CompareEqual(str, maskSlashOrUnderscore);
+                Vector256<sbyte> shift = Avx2.Shuffle(lutShift, Avx2.Add(eq2F, hiNibbles));
+
+                result = Avx2.Add(str, shift);
+
+                return true;
+            }
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public static unsafe int Decode(byte* encodedBytes, ref sbyte decodingMap)
```

Comment:
Can we add comments to these methods that make assumptions about how much data they can read from the arguments? e.g. this method assumes it can read four elements from encodedBytes and that decodingMap has 256 elements.